packages/flutter/lib/src/material/slider.dart

```diff
@@ -389,12 +390,20 @@ class _RenderSlider extends RenderBox {
       ..onTapDown = _handleTapDown
       ..onTapUp = _handleTapUp
       ..onTapCancel = _endInteraction;
-    _reaction = new CurvedAnimation(parent: state.reactionController, curve: Curves.fastOutSlowIn)
-      ..addListener(markNeedsPaint);
+    _reaction = new CurvedAnimation(
+      parent: state.reactionController,
+      curve: Curves.fastOutSlowIn,
+    )..addListener(markNeedsPaint);
     state.enableController.value = isInteractive ? 1.0 : 0.0;
-    _enableAnimation = new CurvedAnimation(parent: state.enableController, curve: Curves.easeInOut)
-      ..addListener(markNeedsPaint);
+    _enableAnimation = new CurvedAnimation(
+      parent: state.enableController,
+      curve: Curves.easeInOut,
+    )..addListener(markNeedsPaint);
     state.positionController.value = _value;
+
+    // Don't need to removeListener on positionController because state has a
+    // 1:1 relationship with this object.
+    state.positionController.addListener(markNeedsPaint);
```

Comment:
this applies to all the addListener calls here, and the situation is more complex than just the one where you directly reference `state`.

Reply:
Hmm, is it because they all have controllers that are part of `state`, and because `state` has a more finite lifetime, then it will be disposed first, and animation controllers don't assert if they still have listeners when disposed. But, (in the normal case) if listeners are detached without calling `removeListener`, then it's possible that the animation controller would still keep calling them after they were detached.  Right?

Reply:
Any call to addListener must be paired with a call to removeListener, or a call to dispose() that implies removeListener (not all objects' dispose methods imply removeListener).

In this case, your render object is created (indirectly) by the state, then the object remains associated with the state (indirectly) until the state is disposed, at which point the render object is never used again. Given this lifecycle (which isn't the guaranteed lifecycle for a render object in general, but happens to be the lifecycle here), you can listen here and be guaranteed that the controllers will be disposed when the render object effectively gets discarded for the last time.

Reply:
Wouldn't it just be safer to add the removeListeners in a detach method for the render object?  I get that it might be slightly more expensive, but if the lifecycle changes and someone doesn't notice it, it's less likely to cause a bug later.